java/ql/src/experimental/Security/CWE/CWE-755/NFEAndroidDoS.ql

```diff
@@ -0,0 +1,107 @@
+  predicate isValueOfMethod(string klass) {
+    this.getMethod().getName() = "valueOf" and
+    this.getQualifier().getType().(RefType).hasQualifiedName("java.lang", klass) and
+    this.getAnArgument().getType().(RefType).hasQualifiedName("java.lang", "String")
+  }
+
+  predicate isParseMethod(string klass, string name) {
+    this.getMethod().getName() = name and
+    this.getQualifier().getType().(RefType).hasQualifiedName("java.lang", klass)
+  }
+
+  predicate throwsNFE() {
+    this.isParseMethod("Byte", "parseByte") or
+    this.isParseMethod("Short", "parseShort") or
+    this.isParseMethod("Integer", "parseInt") or
+    this.isParseMethod("Long", "parseLong") or
+    this.isParseMethod("Float", "parseFloat") or
+    this.isParseMethod("Double", "parseDouble") or
+    this.isParseMethod("Byte", "decode") or
+    this.isParseMethod("Short", "decode") or
+    this.isParseMethod("Integer", "decode") or
+    this.isParseMethod("Long", "decode") or
+    this.isValueOfMethod("Byte") or
+    this.isValueOfMethod("Short") or
+    this.isValueOfMethod("Integer") or
+    this.isValueOfMethod("Long") or
+    this.isValueOfMethod("Float") or
+    this.isValueOfMethod("Double")
+  }
+}
+
+private class SpecialClassInstanceExpr extends ClassInstanceExpr {
+  predicate isStringConstructor(string klass) {
+    this.getType().(RefType).hasQualifiedName("java.lang", klass) and
+    this.getAnArgument().getType().(RefType).hasQualifiedName("java.lang", "String") and
+    this.getNumArgument() = 1
+  }
+
+  predicate throwsNFE() {
+    this.isStringConstructor("Byte") or
+    this.isStringConstructor("Short") or
+    this.isStringConstructor("Integer") or
+    this.isStringConstructor("Long") or
+    this.isStringConstructor("Float") or
+    this.isStringConstructor("Double")
+  }
+}
+
+class NumberFormatException extends RefType {
+  NumberFormatException() { this.hasQualifiedName("java.lang", "NumberFormatException") }
+}
+
+private predicate catchesNFE(TryStmt t) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() instanceof NumberFormatException
+  )
+}
+
+private predicate throwsNFE(Expr e) {
+  e.(SpecialClassInstanceExpr).throwsNFE() or e.(SpecialMethodAccess).throwsNFE()
+}
+
+/**
+ * Taint configuration tracking flow from untrusted inputs to number conversion calls in exported Android compononents.
+ */
+class NFELocalDoSConfiguration extends TaintTracking::Configuration {
+  NFELocalDoSConfiguration() { this = "NFELocalDoSConfiguration" }
+
+  /** Holds if source is a remote flow source */
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  /** Holds if NFE is thrown but not caught */
+  override predicate isSink(DataFlow::Node sink) {
+    exists(Expr e |
+      e.getEnclosingCallable().getDeclaringType() instanceof ExportableAndroidComponent and
```

Comment:
I don't think you check the component actually is exported, only that it can be

Reply:
<code>ExportableAndroidComponent</code> must be exported for the source of <code>RemoteFlowSource</code> in this query thus results are the same regarding Android NFEs. However, explicitly specifying <code>isExported()</code> is unambiguous and helps others to understand so I've made the change.

Reply:
Doh, yes I forgot about that, sorry :) Agree with the increased clarity though, plus it hardens the code against things changing elsewhere.